### PR TITLE
Feature/features.test.js

### PR DESCRIPTION
**Pull request by arcman7**

### CODE REVIEW

File: pysc2/pysc2/lib/actions.js [hunk 1]
```diff
@@ -1,133 +1,134 @@
...
+const path = require('path') //eslint-disable-line
+const s2clientprotocol = require('s2clientprotocol') //eslint-disable-line
+const Enum = require('python-enum') //eslint-disable-line
 const point = require(path.resolve(__dirname, './point.js'))
 const pythonUtils = require(path.resolve(__dirname, './pythonUtils.js'))
 const all_collections_generated_classes = require(path.resolve(__dirname, './all_collections_generated_classes.js'))
 const numpy = require(path.resolve(__dirname, './numpy.js'))
 
-const { spatial_pb, ui_pb } = s2clientprotocol
+const { spatial_pb, ui_pb, common_pb } = s2clientprotocol
```
Comment (jihan1218) [inline pysc2/pysc2/lib/actions.js:9]: Is "common_pb2" instead of "common_pb" ?

Comment (arcman7) [reply pysc2/pysc2/lib/actions.js:9]: yup, the javascript build of the protobuf files doesn't use that "2" suffix

File: pysc2/pysc2/lib/actions.js [hunk 2]
```diff
@@ -136,75 +137,90 @@ function raw_no_op(/*action*/) {
...
-    unit_tags = [unit_tags]
+  const action_cmd = action.getActionRaw().getUnitCommand()
+  action_cmd.setAbilityId(ability_id)
+  action_cmd.setQueueCommand(queued)
+  if (isinstance(unit_tags, [Array])) {
+    unit_tags.forEach((unit_tag) => {
+      action_cmd.addUnitTags(unit_tag)
+    })
+  } else {
+    action_cmd.addUnitTags(unit_tags)
```
Comment (jihan1218) [inline pysc2/pysc2/lib/actions.js:154]: If unit_tags is not array, then do we need to change it to array?

Comment (arcman7) [reply pysc2/pysc2/lib/actions.js:154]: no, look at what the code does
if array 
  add each tag elementwise 
else
 add single tag

Comment (arcman7) [reply pysc2/pysc2/lib/actions.js:154]: In the python code they just wanted to make use of List.extends

File: pysc2/pysc2/lib/features.js [hunk 3]
```diff
@@ -1029,11 +1020,11 @@ class Features {
...
-    this.available_actions = sw.decorate(this.available_actions)
-    this.transform_action = sw.decorate(this.transform_action)
-    this.reverse_action = sw.decorate(this.reverse_action)
-    this.reverse_raw_action = sw.decorate(this.reverse_raw_action)
+    this.transform_obs = sw.decorate(this.transform_obs.bind(this))
+    this.available_actions = sw.decorate(this.available_actions.bind(this))
+    this.transform_action = sw.decorate(this.transform_action.bind(this))
+    this.reverse_action = sw.decorate(this.reverse_action.bind(this))
+    this.reverse_raw_action = sw.decorate(this.reverse_raw_action.bind(this))
   }
```
Comment (jihan1218) [inline pysc2/pysc2/lib/features.js:1028]: why is this decorator used here?

File: pysc2/pysc2/lib/features.js [hunk 4]
```diff
@@ -1060,34 +1051,34 @@ class Features {
...
+    const world_tl_to_feature_minimap = new transform.Linear(scale)
+    this._world_to_minimap_px = new transform.Chain(
       this._world_to_world_tl,
       world_tl_to_feature_minimap,
-      transform.PixelToCoord()
+      new transform.PixelToCoord()
     )
     this._camera_size = (
-      raw_resolution / (map_size.max_dim() * camera_width_world_units)
+      raw_resolution.div(map_size.max_dim() * (camera_width_world_units))
```
Comment (jihan1218) [inline pysc2/pysc2/lib/features.js:1081]: It should be 
raw_resolution.div(map_size.max_dim())* (camera_width_world_units)
original: 
raw_resolution / map_size.max_dim() * camera_width_world_units)

File: pysc2/pysc2/lib/pythonUtils.js [hunk 5]
```diff
@@ -238,6 +438,7 @@ module.exports = {
   map,
   NotImplementedError,
   randomUniform,
+  setUpProtoAction,
   String,
   sum,
   ValueError,
```
Comment (jihan1218) [inline pysc2/pysc2/lib/pythonUtils.js:444]: We are missing "pythonWith" function

File: pysc2/pysc2/lib/features.js [hunk 6]
```diff
@@ -1500,6 +1494,7 @@ class Features {
     raw = obs.observation.raw_data
 
     if (aif.use_feature_units) {
+      console.log('*** THIS SECTION MUST BE FIXED ****')
```
Comment (jihan1218) [inline pysc2/pysc2/lib/features.js:1497]: Is this a note?

Comment (arcman7) [reply pysc2/pysc2/lib/features.js:1497]: yes

**Pull request merged**